### DISCUSSION TITLE
External Authentication and support for third party JWT

### DISCUSSION THREAD

**Discussion by CallMeLaNN** [directus/directus#7442] [Feature Requests]
This is my first post on a very good community. I've been trying Directus for sometime after realizing the existence of this headless CMS alternatives with more useful features. I heard about the journey moving to NodeJS, in the code I saw it using Vite, some other nice libraries and clean UI, you guys are awesome.

## Authentication

Supporting for JWT, OAuth and OIDC is a good approach. After using it, checking issues and discussions, there are a little bit use cases that we can tackle. My proposal basically in three main areas that we may improve with less invasive way.

#### Proposals:

1. Third party JWT: Validate and recognize token from external AS or IdP.
2. OAuth and OIDC: Some settings and integration improvement.
3. User schema: Handling unique ID, email, profile and mapping to external ID.

My use cases similar with other reported concerns in this areas. I'll prioritize (1) and the rest can be discussed later.

## Third party JWT

This is most discussed topic for authentication. I managed to do this in a custom API endpoint. However consider this (so I'll explain it well to the other readers).

#### Considerations:

1. I feel like this is best to be built-in instead of custom API.
1. Things like some code sharing, I have been duplicating it from Directus api.
1. Automatically fetch email, first name and last name from OAuth User Info endpoint just like `grant` did.
1. Automatically create the user record and assign default role (for the first time) after authorized.
1. Better and native for mobile dev or user than using OAuth flow in an in-app browser.
1. Mobile native login page requirement. Resource Owner Password Credentials flow is not recommended anymore.
1. We can use registered claims from RFC7519 that suggest `sub` to identify user. Optionally issuer `iss` and audience `aud` as an extra validation for limiting the usage.
1. Sharing/deploying JWT secret or using public key/JWK to verify signature is not a problem.
1. Some libraries can help us to fetch public key from remote JWKS and verify easily.
1. I've seen and use this pattern on some backends supporting external AS like Synapse matrix.org.
1. Expiry and refreshing token shouldn't be an issue, it is up to the client to keep track of tokens.

That is long enough 😃, I'm willing and ready to contribute if this is convincing enough. Two approaches, we can have either one or both.

#### Solutions:

1. Direct: Recognizing external access token directly to give access.
  Update existing JWT verification and add some settings. It does verify external JWT but just to add (3) and (4) above and also make claim [`id`](https://github.com/directus/directus/blob/eaa2f47453e1ab978443703165ed8ebe4caf149e/api/src/middleware/authenticate.ts#L46)/`sub` configurable.
2. STS: Using Security Token Service concept that exchange external access token with internal one.
  Add one API like `POST /auth/loginWithJWT` and the same like direct approach.

The (2) is what I did but from the custom API endpoint. Anyway, each of this serve for different cases to become client, server or both of AS/IdP. It is easier to move forward with (2) first.

#### Expectation:

To make use of this, just add extra env variables and call the API.

I would love to hear some feedbacks.

**Comment by rijkvanzanten** [comment]
I like your thinking here! Does RFC7519 also go into depth on where to pull the original profile from? 

We should also discuss how this ties into user _creation_ when somebody uses a token that doesn't match to an existing user yet 🤔

**Reply by rijkvanzanten** [reply]
Also, would you use this external JWT token to retrieve a Directus one once, or would you envision this external JWT be used for every request? If the latter, it might add quite a bit of overhead in checks 🤔

**Reply by CallMeLaNN** [reply]
> Does RFC7519 also go into depth on where to pull the original profile from?

Nope, It only define standard claims that is optional other than plain JWT. At least it standardize identifier, how to verify it and also check if the token coming from configured issuer.

OIDC is the one defined the profile on top of OAuth in two ways. Either from OAuth token endpoint as [`id_token`](https://auth0.com/docs/tokens/id-tokens/id-token-structure) (but not all provider put profile in it) or from OIDC **user info endpoint** just like what `grant` did using `OAUTH_*_PROFILE_URL`. This endpoint along with other OAuth endpoints can also be discovered [like this](https://accounts.google.com/.well-known/openid-configuration). So imagine in .env we can set to turn on discovery or to manually specify profile URL.

> how this ties into user creation when somebody uses a token that doesn't match to an existing user yet 🤔

I was thinking why it doesn't happen after authorized from `grant` 😅. Seems like to have a hook to complete the OAuth config. In my case for external JWT, my `/custom/auth/loginFromJwt` do it manually.

Fetching external user profile endpoint should be done once if the user doesn't exist yet. How to match the user is a good question that I will combine similar issues in `3. User schema` for later discussion to complete this. Basically external JWT contain user id that we can use to map these internal and external user id.

In the case of social login, to match based on email is interesting topic, but it does make sense to match just based on social user id because once the map is there, no need to know the email after the creation, hence no more user profile lookup. In this case, it will cover most SSO for JWT, OAuth social and OIDC cases.

If you think this is a lot, we can simplify this by open up the flexibility of user table for id: string and optional email. Think like it is just a cache of real user record somewhere else but only for one provider and at the same time having local users with email.

> would you envision this external JWT be used for every request? If the latter, it might add quite a bit of overhead in checks 🤔

It depends on if we implement direct use or exchange then token like the solutions in OP. Exchange (like STS) is only once per Directus access token lifetime, like 15m or 1hr. Even though we do the direct use of external JWT, IMO there is no difference verifying internal or external token, both will load Directus user/role record and share the existing `authenticate` method.

**Reply by rijkvanzanten** [reply]
Thanks for the in depth information!

This all makes sense to me. The main difficulty here is trying to figure out how to inject any custom logic that's responsible for validating the JWT. For OIDC/oAuth/LDAP, that's something we can do natively in the system. For a custom JWT, there needs to be a custom bit of JS that will read the jwt, and act accordingly with a custom third party service to refresh/check the token 🤔  This could theoretically be handled by a hook, as long as we have a "validate JWT" event you can tap in to

**Reply by dalotodo** [reply]
Hello, everyone,

This is my first discussion on your repo, so first of all I would like to thank you for such a good app!!

Regarding this topic itself, after a deep review of the repo code I would like to suggest a very basic approach, pretty much compatible with your actual development.

First of all, when talking about JWT validation we must have in mind that within SPAs, retrieved access tokens from third parties **are actually valid** to use APIs from those parties.

So, in short: 3rd party access tokens are valid, and you can check them agains their APIs. 

So, what if you just exchange that access_token against directus, and if its valid, retrieve a Directus-signed access token?

So the flow would be like this:

- User signs-in into the Application (SPA) by using 3rd party auth libraries (MSAL, Auth0, etc.) --> These libraries take care of the SPA auth flow and token validness
- SPA retrieves a valid access_token from the 3rd party
- SPA _exchanges_ that token agains Directus API backend (DIRECTUS)
- DIRECTUS checks that token against a _get-profile-endpoint_ from the issuing party
- If successful, DIRECTUS should be able to retrieve user profile information from 3rd party
- DIRECTUS adds a single-use refresh_token into the `directus_sessions` collection, with very short lifetime
- DIRECTUS calls the refresh() method of the AuthenticationService, to get back to the normal application flow
- DIRECTUS returns to the SPA the _refreshed_ tokens from the previous call
- SPA can then use it to either store it in the `auth.storage` of the SDK, or exchange it for a new token

Things to have in mind:
- This method is fully compatible with existing application flow
- This method is extensible to any available authentication service (Google, Facebook, Auth0, etc.) as long as the party has a public endpoint for token validation or profile retrieval
- To avoid security issues with token exchange (i.e. Directus makes delegated calls to the endpoint), developers and sys-admins have the ability to narrow the scopes of the issued token, to be only valid for profile queries

Below i've left an example extension endpoint draft that performs the described scenario against Microsoft Azure (Graph API).

Hope it helps!

Best.

```typescript
import { defineEndpoint } from '@directus/extensions-sdk';
import axios from 'axios'
import { addSeconds } from 'date-fns'
import { nanoid } from 'nanoid'

const MICROSOFT_GRAPH_URL = 'https://graph.microsoft.com/v1.0'

type ODataMetaInfo = {
	'@odata.context'?: string;
	'@odata.count'?: number;
	'@odata.nextLink'?: string;
}
type ODataResponse<T> = ODataMetaInfo & T;

type GraphUser = {
	givenName: string;
	surname: string;
	displayName: string;
	userPrincipalName: string;
	mail: string;
}

class AzureGraphService {
	private readonly _accessToken: string;	

	constructor(accessToken: string) {
		this._accessToken = accessToken;		
	}

	public async me() {
		const url = `${MICROSOFT_GRAPH_URL}/me`
		
		const response = await axios.get<ODataResponse<GraphUser>>(url, {
			headers: {
				'Authorization': `Bearer ${this._accessToken}`
			}
		});

		return response.data;
	}
}


type BodyWithAccessToken = { accessToken?: string; };

export default defineEndpoint({
	id: 'oauth2',
	handler: (router, context) => {
		const { services, exceptions, logger, database, getSchema } = context;

		// Endpoint for ping (checking the extension is up)
		router.get('/azure', async (req, res, next) => {
			return res.json({ message: 'Me want cookie!' })
		})

		/**
		 * Authenticate agains Microsoft Graph API (Azure)
		 */
		router.post('/azure', async (req, res, next) => {
			const { ForbiddenException } = exceptions;
			const { AuthenticationService, UsersService } = services;			

			const requestBody = req.body as BodyWithAccessToken;

			// Get Azure Access token previously retrieved from SPA
			const azureAccessToken = requestBody?.accessToken;
			if (!azureAccessToken) {
				return next(new ForbiddenException());
			}

			// Try to retrieve information from the Microsoft Graph Endpoint
			let profile: GraphUser | undefined;
			const graph = new AzureGraphService(azureAccessToken)
			try {
				profile = await graph.me()
			} catch (err) {
				logger.error(err)
			}
			
			if (!profile) {
				return next(new ForbiddenException());
			}

			const schema = await getSchema();

			// Find user 
			const foundUser = await database
				.select('id', 'status', 'password', 'token')
				.from('directus_users')
				.whereRaw('LOWER(??) = ?', ['email', profile.mail])
				.first();

			if (!foundUser) {
				return next(new ForbiddenException());
			}

			const usersService = new UsersService({ schema, knex: database });			

			// Create one time session token (OTST) with 30s lifetime
			const otst = nanoid(64); 
			const expiration = addSeconds(new Date(), 30)
			await database.insert({ token: otst, user: foundUser.id, expires: expiration, ip: req.ip, user_agent: req.get('user-agent') })
				.into('directus_sessions')
				;		

			// Try to authenticate by refreshing (OTST)
			const accountability = {
				ip: req.ip,
				userAgent: req.get('user-agent'),
				role: null,
			};
			const authenticationService = new AuthenticationService({
				accountability: accountability,
				schema: schema
			});			

			try {				
				// Refresh token
				const { accessToken, refreshToken, expires } = await authenticationService.refresh( otst );				
				const payload = {
					data: { 
						access_token: accessToken, 
						refresh_token: refreshToken,
						expires 
					},
				} as Record<string, Record<string, any>>;		
				return res.json( payload )				
			}
			catch (err) {
				logger.error(err);
				return next(new ForbiddenException());
			}			
		});
	}
})



```

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
